optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java

```diff
@@ -0,0 +1,158 @@
+ */
+
+package org.optaplanner.benchmark.impl.statistic;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.api.solver.event.BestSolutionChangedEvent;
+import org.optaplanner.core.api.solver.event.SolverEventListener;
+import org.optaplanner.core.config.solver.metric.SolverMetric;
+import org.optaplanner.core.impl.phase.event.PhaseLifecycleListener;
+import org.optaplanner.core.impl.phase.scope.AbstractPhaseScope;
+import org.optaplanner.core.impl.phase.scope.AbstractStepScope;
+import org.optaplanner.core.impl.score.definition.ScoreDefinition;
+import org.optaplanner.core.impl.solver.DefaultSolver;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.search.Search;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+public class StatisticRegistry extends SimpleMeterRegistry implements PhaseLifecycleListener, SolverEventListener {
+
+    List<Consumer<Long>> stepMeterListenerList = new ArrayList<>();
+    List<Consumer<Long>> bestSolutionMeterListenerList = new ArrayList<>();
+    long bestSolutionChangedTimestamp = Long.MIN_VALUE;
+    ScoreDefinition scoreDefinition;
+
+    public StatisticRegistry(DefaultSolver solver) {
+        scoreDefinition = solver.getSolverScope().getScoreDefinition();
+    }
+
+    public void addListener(SolverMetric metric, Consumer<Long> listener) {
+        if (isMetricBestSolutionBased(metric)) {
+            bestSolutionMeterListenerList.add(listener);
+        } else {
+            stepMeterListenerList.add(listener);
+        }
+    }
+
+    private boolean isMetricBestSolutionBased(SolverMetric metric) {
+        switch (metric) {
+            case BEST_SCORE:
+            case BEST_SOLUTION_MUTATION:
+            case CONSTRAINT_MATCH_TOTAL_BEST_SCORE:
+            case PICKED_MOVE_TYPE_BEST_SCORE_DIFF:
+                return true;
+            case MEMORY_USE:
+            case STEP_SCORE:
+            case ERROR_COUNT:
+            case SOLVE_LENGTH:
+            case MOVE_COUNT_PER_STEP:
+            case SCORE_CALCULATION_SPEED:
+            case PICKED_MOVE_TYPE_STEP_SCORE_DIFF:
+            case CONSTRAINT_MATCH_TOTAL_STEP_SCORE:
+                return false;
+            default:
+                throw new IllegalStateException("SolverMetric (" + metric + ") does not have a case");
+        }
+    }
+
+    public Set<Meter.Id> getMeterIds(SolverMetric metric, Tags runId) {
+        return Search.in(this).name(name -> name.startsWith(metric.getMeterId())).tags(runId)
+                .meters().stream().map(meter -> meter.getId())
+                .collect(Collectors.toSet());
```

Comment:
In this fluent chain, I find it hard to see where micrometer ends and streams start.
May be worth splitting former into its own variable.
Also, `Meter::getId`.